core/src/main/scala/cats/data/NonEmptySet.scala

```diff
@@ -428,7 +428,7 @@ sealed abstract private[data] class NonEmptySetInstances extends NonEmptySetInst
 }
 
 sealed abstract private[data] class NonEmptySetInstances0 extends NonEmptySetInstances1 {
-  implicit def catsDataHashForNonEmptySet[A: Order: Hash]: Hash[NonEmptySet[A]] =
+  implicit def catsDataHashForNonEmptySet[A: Hash]: Hash[NonEmptySet[A]] =
```

Comment:
Can someone explain (or guess at least) why does this change not concern Mima? I removed the unused implicit `Order` parameter but seems that Mima is completely fine about that.

Reply:
Um, that seems suspect. Let me look into it, hopefully there's a good explanation ...

Reply:
Ok, I'm really confused ... but is this all private??

`NonEmptySetInstances0` is private, and extended by `NonEmptySetInstances`, which is private, which is implemented by `NonEmptySetImpl`, also private.

Reply:
Right.. But the implicit method itself is public and should be publicly accessible as a part of the `NonEmptySet` object API.

Reply:
That's the thing .. AFAICT there _is_ no (public) `object NonEmptySet` anywhere in Cats 🙃

Reply:
That's correct... It is just a val: `val NonEmptySet = NonEmptySetImpt`...

Reply:
..and the same trick is used for `NonEmptyMap` and `NonEmptyChain`

Reply:
Ah, thanks, I'm sorry. Didn't realize that's how it's implemented.
https://github.com/typelevel/cats/blob/708cbaba2e6ac07a5ac1819b6c18b0c83a2faaba/core/src/main/scala/cats/data/package.scala#L39

The bad news is, we won't get any MiMa support on this :( The reason is that the type of `NonEmptySet` is `NonEmptySetImpl.type` aka `NonEmptySetImpl$`, which is package-private, so MiMa ignores it. Similarly, you won't find any of these APIs in the scaladocs.

In fact, I'm pretty surprised this works at all. This smells a lot to me like https://github.com/scala/bug/issues/6794.

Reply:
If we want to get MiMa support for these, we'll have to expose them as a public `class` or `trait` somewhere, which should match the type of that `val NonEmptySet`.

Reply:
Oh, I see... Is it even possible to configure Mima to force it looking into some specific objects even if they're package private? We would need to add those 3 objects then at least...

Reply:
You can open an issue on MiMa, but not sure what can be done there.

https://github.com/typelevel/cats/blob/708cbaba2e6ac07a5ac1819b6c18b0c83a2faaba/core/src/main/scala/cats/data/NonEmptySet.scala#L30

Instead I propose this solution: 
```scala
sealed class NonEmptySetCompanion extends NonEmptySetInstances with Newtype`
private[data] object NonEmptySetImpl extends NonEmptySetCompanion

val NonEmptySet: NonEmptySetCompanion = NonEmptySetImpl
```

I think that should get us what we want.

Reply:
Ah, but note, even if we do my proposal and publish it as 2.8.0, it won't check bincompat for us against 2.7.0 or earlier. For that we would really need changes in MiMa.

Reply:
Does Mima check against previous snapshots created out of the current `main` branch? What if I first create a PR with public sealed class as you proposed, then we merge it and then – I wonder – will Mima be checking against the already merged code, wdyt?

Reply:
And another question... Should it be really a class? Or will a public sealed trait be enough there as well?

Reply:
> Does Mima check against previous snapshots created out of the current main branch?

No, it's not currently configured to check snapshots, but in theory we could try and configure that. Technically, we don't guarantee binary-compatibility against snapshots, and it might be annoying if we add something but then remove it or change it before the release.

> And another question... Should it be really a class? Or will a public sealed trait be enough there as well?

Yeah, `class`, `trait`, whatever :) so long as its public.

Reply:
Btw, there is one other option. Cats has a manual bincompat test here:
https://github.com/typelevel/cats/blob/main/binCompatTest/src/main/scala-2/catsBC/MimaExceptions.scala

It actually compiles against an old Cats artifact, and then attempts to run it with the current branch. So that is the true test for bincompat (unlike MiMa, which is just doing an analysis). So in theory we could add use all of these methods in there, and it will be tested as part of CI even if MiMa can't check it for us.

Reply:
Oh, cool, I didn't know about it. So do you mean – if I add an appropriate check for that implicit there, then it will be checked against the previous published artifact and we can see if really BinCompat or not – is that correct?

Reply:
Yes, that's right. If you do something which uses `catsDataHashForNonEmptySet` it _should_ fail since you removed the `Order`.

Reply:
You could possibly try a bytecode level compatibility checker like [Revapi](https://revapi.org/revapi-site/main/getting-started.html) or [SigTest](https://wiki.openjdk.java.net/display/CodeTools/SigTest) as a one off, although I don't know how well they will work on Scala projects

Reply:
Hopefully handled in #4219.